File: src/test/java/name/gyger/jmoney/report/ReportServiceTests.java
package name.gyger.jmoney.report;

import name.gyger.jmoney.EntityFactoryKt;
import name.gyger.jmoney.account.AccountService;
import name.gyger.jmoney.account.Entry;
import name.gyger.jmoney.account.EntryService;
import name.gyger.jmoney.category.Category;
import name.gyger.jmoney.category.CategoryService;
import name.gyger.jmoney.session.SessionService;
import name.gyger.jmoney.util.DateUtilKt;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.transaction.Transactional;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@SpringBootTest
@Transactional
public class ReportServiceTests {

    @Autowired
    ReportService reportService;

    @Autowired
    CategoryService categoryService;

    @Autowired
    AccountService accountService;

    @Autowired
    EntryService entryService;

    @Autowired
    SessionService sessionService;

    @PersistenceContext
    private EntityManager em;

    @Before
    public void setup() {
        sessionService.initSession();
        em.flush();
        em.clear();
    }

    @Test
    public void testBalances() {
        long accountA = EntityFactoryKt.createAccount("A", 0, accountService);
        long accountB = EntityFactoryKt.createAccount("B", 0, accountService);

        Date date = DateUtilKt.parse("2000-01-01");
        assertThat(reportService.getBalances(date)).hasSize(3);

        Entry entry = new Entry();
        entry.setAccountId(accountA);
        entry.setDate(date);
        entry.setAmount(15);
        EntityFactoryKt.createEntries(entry, 10, entryService);

        entry.setAccountId(accountB);
        entry.setAmount(-5);
        EntityFactoryKt.createEntries(entry, 10, entryService);

        List<Balance> balances = reportService.getBalances(date);
        assertThat(balances).hasSize(3);

        Balance balanceA = balances.get(0);
        assertThat(balanceA.getBalance()).isEqualTo(150);
        assertThat(balanceA.isTotal()).isFalse();
        assertThat(balanceA.getAccountName()).isEqualTo("A");

        Balance balanceB = balances.get(1);
        assertThat(balanceB.getBalance()).isEqualTo(-50);
        assertThat(balanceB.isTotal()).isFalse();
        assertThat(balanceB.getAccountName()).isEqualTo("B");

        Balance balanceTotal = balances.get(2);
        assertThat(balanceTotal.getBalance()).isEqualTo(100);
        assertThat(balanceTotal.isTotal()).isTrue();
        assertThat(balanceTotal.getAccountName()).isEqualTo("Gesamt");
    }

    @Test
    public void testCashFlow() {
        long accId = EntityFactoryKt.createAccount("A", 0, accountService);

        long catA = EntityFactoryKt.createTopLevelCategory("A", categoryService);
        long catB = EntityFactoryKt.createTopLevelCategory("B", categoryService);
        long catC = EntityFactoryKt.createTopLevelCategory("C", categoryService);
        long catD = EntityFactoryKt.createTopLevelCategory("D", categoryService);

        Date date = DateUtilKt.parse("2000-10-07");
        Entry entry = new Entry();
        entry.setAccountId(accId);
        entry.setDate(date);
        entry.setAmount(7);
        entry.setCategoryId(catA);
        EntityFactoryKt.createEntries(entry, 8, entryService);

        entry.setAmount(-9);
        entry.setCategoryId(catB);
        EntityFactoryKt.createEntries(entry, 10, entryService);

        entry.setAmount(11);
        entry.setCategoryId(catC);
        EntityFactoryKt.createEntries(entry, 12, entryService);

        em.flush();
        em.clear();

        Date from = DateUtilKt.parse("2000-01-01");
        Date to = DateUtilKt.parse("2001-01-01");
        List<CashFlow> cashFlow = reportService.getCashFlow(from, to);
        assertThat(cashFlow).hasSize(7);

        CashFlow cashFlowA = cashFlow.get(0);
        assertThat(cashFlowA.getIncome()).isEqualTo(56);
        assertThat(cashFlowA.getExpense()).isNull();
        assertThat(cashFlowA.getDifference()).isNull();
        assertThat(cashFlowA.isTotal()).isFalse();

        CashFlow cashFlowB = cashFlow.get(2);
        assertThat(cashFlowB.getIncome()).isNull();
        assertThat(cashFlowB.getExpense()).isEqualTo(90);
        assertThat(cashFlowB.getDifference()).isNull();
        assertThat(cashFlowB.isTotal()).isFalse();

        CashFlow cashFlowC = cashFlow.get(4);
        assertThat(cashFlowC.getIncome()).isEqualTo(132);
        assertThat(cashFlowC.getExpense()).isNull();
        assertThat(cashFlowC.getDifference()).isNull();
        assertThat(cashFlowC.isTotal()).isFalse();

        CashFlow cashFlowTotal = cashFlow.get(6);
        assertThat(cashFlowTotal.getIncome()).isEqualTo(188);
        assertThat(cashFlowTotal.getExpense()).isEqualTo(90);
        assertThat(cashFlowTotal.getDifference()).isEqualTo(98);
        assertThat(cashFlowTotal.isTotal()).isTrue();
    }

    @Test
    public void testInconsistentSplitEntry() {
        long accountId = EntityFactoryKt.createAccount("my account", 0, accountService);

        Category split = categoryService.getSplitCategory();
        Entry entry = new Entry();
        entry.setCategoryId(split.getId());
        entry.setAccountId(accountId);
        entry.setAmount(12715);
        entryService.createEntry(entry);
        assertThat(reportService.getInconsistentSplitEntries()).hasSize(1);

        Entry subEntry = new Entry();
        subEntry.setAmount(12715);
        List<Entry> subEntries = new ArrayList<>();
        subEntries.add(subEntry);
        entry.setSubEntries(subEntries);
        entryService.updateEntry(entry);
        assertThat(reportService.getInconsistentSplitEntries()).hasSize(0);
    }

    @Test
    public void testEntriesWithoutCategory() {
        long accountId = EntityFactoryKt.createAccount("my account", 0, accountService);
        Entry entryDto = new Entry();
        entryDto.setAccountId(accountId);
        EntityFactoryKt.createEntries(entryDto, 37, entryService);
        assertThat(reportService.getEntriesWithoutCategory()).hasSize(37);
    }

    @Test
    public void testEntriesForCategory() {
        Category newCat = new Category();
        newCat.setName("NEW");
        newCat.setParentId(categoryService.getRootCategory().getId());
        long newCatId = categoryService.createCategory(newCat);

        long accountId = EntityFactoryKt.createAccount("my account", 0, accountService);
        Date date = DateUtilKt.parse("2000-10-07");
        Entry entry = new Entry();
        entry.setAccountId(accountId);
        entry.setDate(date);
        entry.setCategoryId(newCatId);
        EntityFactoryKt.createEntries(entry, 7, entryService);

        assertThat(reportService.getEntriesForCategory(newCatId, date, date)).hasSize(7);
    }

}

File: src/test/kotlin/name/gyger/jmoney/report/ReportServiceTests.kt
package name.gyger.jmoney.report

import name.gyger.jmoney.account.AccountService
import name.gyger.jmoney.account.Entry
import name.gyger.jmoney.account.EntryService
import name.gyger.jmoney.category.Category
import name.gyger.jmoney.category.CategoryService
import name.gyger.jmoney.createAccount
import name.gyger.jmoney.createEntries
import name.gyger.jmoney.createTopLevelCategory
import name.gyger.jmoney.session.SessionService
import name.gyger.jmoney.util.parse
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit4.SpringRunner
import java.util.*
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext
import javax.transaction.Transactional

@RunWith(SpringRunner::class)
@SpringBootTest
@Transactional
open class ReportServiceTests {

    @Autowired
    lateinit var reportService: ReportService

    @Autowired
    lateinit var categoryService: CategoryService

    @Autowired
    lateinit var accountService: AccountService

    @Autowired
    lateinit var entryService: EntryService

    @Autowired
    lateinit var sessionService: SessionService

    @PersistenceContext
    lateinit private var em: EntityManager

    @Before
    fun setup() {
        sessionService.initSession()
        em.flush()
        em.clear()
    }

    @Test
    fun testBalances() {
        val accountA = createAccount("A", 0, accountService)
        val accountB = createAccount("B", 0, accountService)

        val date = parse("2000-01-01")
        assertThat(reportService.getBalances(date)).hasSize(3)

        val entry = Entry()
        entry.accountId = accountA
        entry.date = date
        entry.amount = 15
        createEntries(entry, 10, entryService)

        entry.accountId = accountB
        entry.amount = -5
        createEntries(entry, 10, entryService)

        val balances = reportService.getBalances(date)
        assertThat(balances).hasSize(3)

        val balanceA = balances[0]
        assertThat(balanceA.balance).isEqualTo(150)
        assertThat(balanceA.isTotal).isFalse()
        assertThat(balanceA.accountName).isEqualTo("A")

        val balanceB = balances[1]
        assertThat(balanceB.balance).isEqualTo(-50)
        assertThat(balanceB.isTotal).isFalse()
        assertThat(balanceB.accountName).isEqualTo("B")

        val balanceTotal = balances[2]
        assertThat(balanceTotal.balance).isEqualTo(100)
        assertThat(balanceTotal.isTotal).isTrue()
        assertThat(balanceTotal.accountName).isEqualTo("Total")
    }

    @Test
    fun testCashFlow() {
        val accId = createAccount("A", 0, accountService)

        val catA = createTopLevelCategory("A", categoryService)
        val catB = createTopLevelCategory("B", categoryService)
        val catC = createTopLevelCategory("C", categoryService)

        val date = parse("2000-10-07")
        val entry = Entry()
        entry.accountId = accId
        entry.date = date
        entry.amount = 7
        entry.categoryId = catA
        createEntries(entry, 8, entryService)

        entry.amount = -9
        entry.categoryId = catB
        createEntries(entry, 10, entryService)

        entry.amount = 11
        entry.categoryId = catC
        createEntries(entry, 12, entryService)

        em.flush()
        em.clear()

        val from = parse("2000-01-01")
        val to = parse("2001-01-01")
        val cashFlow = reportService.getCashFlow(from, to)
        assertThat(cashFlow).hasSize(7)

        val cashFlowA = cashFlow[0]
        assertThat(cashFlowA.income).isEqualTo(56)
        assertThat(cashFlowA.expense).isNull()
        assertThat(cashFlowA.difference).isNull()
        assertThat(cashFlowA.isTotal).isFalse()

        val cashFlowB = cashFlow[2]
        assertThat(cashFlowB.income).isNull()
        assertThat(cashFlowB.expense).isEqualTo(90)
        assertThat(cashFlowB.difference).isNull()
        assertThat(cashFlowB.isTotal).isFalse()

        val cashFlowC = cashFlow[4]
        assertThat(cashFlowC.income).isEqualTo(132)
        assertThat(cashFlowC.expense).isNull()
        assertThat(cashFlowC.difference).isNull()
        assertThat(cashFlowC.isTotal).isFalse()

        val cashFlowTotal = cashFlow[6]
        assertThat(cashFlowTotal.income).isEqualTo(188)
        assertThat(cashFlowTotal.expense).isEqualTo(90)
        assertThat(cashFlowTotal.difference).isEqualTo(98)
        assertThat(cashFlowTotal.isTotal).isTrue()
    }

    @Test
    fun testInconsistentSplitEntry() {
        val accountId = createAccount("my account", 0, accountService)

        val split = categoryService.getSplitCategory()
        val entry = Entry()
        entry.categoryId = split.id
        entry.accountId = accountId
        entry.amount = 12715
        entryService.deepSave(entry)
        assertThat(reportService.getInconsistentSplitEntries()).hasSize(1)

        val subEntry = Entry()
        subEntry.amount = 12715
        val subEntries = ArrayList<Entry>()
        subEntries.add(subEntry)
        entry.subEntries = subEntries
        entryService.deepSave(entry)
        assertThat(reportService.getInconsistentSplitEntries()).hasSize(0)
    }

    @Test
    fun testEntriesWithoutCategory() {
        val accountId = createAccount("my account", 0, accountService)
        val entryDto = Entry()
        entryDto.accountId = accountId
        createEntries(entryDto, 37, entryService)
        assertThat(reportService.getEntriesWithoutCategory()).hasSize(37)
    }

    @Test
    fun testEntriesForCategory() {
        val newCat = Category()
        newCat.name = "NEW"
        newCat.parentId = categoryService.getRootCategory().id
        val newCatId = categoryService.createCategory(newCat)

        val accountId = createAccount("my account", 0, accountService)
        val date = parse("2000-10-07")
        val entry = Entry()
        entry.accountId = accountId
        entry.date = date
        entry.categoryId = newCatId
        createEntries(entry, 7, entryService)

        assertThat(reportService.getEntriesForCategory(newCatId, date, date)).hasSize(7)
    }

}
